rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -69,21 +69,44 @@
 
 namespace
 {
+//==============================================================================
+// Helper function to create a copy of a string after removing any
+// leading or trailing slashes.
+std::string strip_slashes(const char * const str)
+{
+  std::string ret = std::string(str);
+  const std::size_t len = strlen(str);
+  std::size_t start = 0;
+  std::size_t end = len - 1;
+  if (str[0] == '/') {
+    ++start;
+  }
+  if (str[end] == '/') {
+    --end;
+  }
+  return ret.substr(start, end - start);
+}
 
 //==============================================================================
-// A function to take ros topic names and convert them to valid Zenoh keys.
+// A function that generates a key expression for message transport of the format
+// <ros_domain_id>/<topic_name>/<topic_type>/<topic_hash>
 // In particular, Zenoh keys cannot start or end with a /, so this function
 // will strip them out.
-// The Zenoh key is also prefixed with the ros_domain_id.
 // Performance note: at present, this function allocates a new string and copies
 // the old string into it. If this becomes a performance problem, we could consider
 // modifying the topic_name in place. But this means we need to be much more
 // careful about who owns the string.
-z_owned_keyexpr_t ros_topic_name_to_zenoh_key(const char * const topic_name, size_t domain_id)
+z_owned_keyexpr_t ros_topic_name_to_zenoh_key(
+  const std::size_t domain_id,
+  const char * const topic_name,
+  const char * const topic_type,
+  const char * const type_hash)
 {
-  const std::string keyexpr_str = std::to_string(domain_id) +
-    "/" +
-    rmw_zenoh_cpp::liveliness::mangle_name(topic_name);
+  const std::string keyexpr_str = std::to_string(domain_id) + "/" +
+    strip_slashes(topic_name) + "/" +
+    std::string(topic_type) + "/" +
+    std::string(type_hash);
```

Comment:
We can avoid at least one allocation here by doing:

```
  std::string keyexpr_str = std::to_string(domain_id);
  keyexpr_str += "/";
  keyexpr_str += strip_slashes(topic_name);
  keyexpr_str += "/";
  keyexpr_str += topic_type;
  keyexpr_str += "/";
  keyexpr_str += type_hash;
```

Reply:
Thanks for the feedback. Done https://github.com/ros2/rmw_zenoh/pull/171/commits/50d7339473b4c80e328d69f1b2cbd6e096e0a02f